File: runtime/ios/Sources/GSXNativeKit/GSXAuth.swift
import Foundation

public struct GSXAuthExchangeRequest: Codable, Equatable {
    public var strategy: String
    public var credentials: [String: String]
    public var attributes: [String: String]

    public init(
        strategy: String,
        credentials: [String: String] = [:],
        attributes: [String: String] = [:]
    ) {
        self.strategy = strategy
        self.credentials = credentials
        self.attributes = attributes
    }
}

public struct GSXAuthTokenResponse: Codable, Equatable {
    public var accessToken: String
    public var refreshToken: String?
    public var expiresInSeconds: Int?
    public var tokenType: String

    public init(
        accessToken: String,
        refreshToken: String? = nil,
        expiresInSeconds: Int? = nil,
        tokenType: String = "Bearer"
    ) {
        self.accessToken = accessToken
        self.refreshToken = refreshToken
        self.expiresInSeconds = expiresInSeconds
        self.tokenType = tokenType
    }

    private enum CodingKeys: String, CodingKey {
        case accessToken = "access_token"
        case refreshToken = "refresh_token"
        case expiresInSeconds = "expires_in"
        case tokenType = "token_type"
    }

    public init(from decoder: Decoder) throws {
        let container = try decoder.container(keyedBy: CodingKeys.self)
        accessToken = try container.decode(String.self, forKey: .accessToken)
        refreshToken = try container.decodeIfPresent(String.self, forKey: .refreshToken)
        expiresInSeconds = try container.decodeIfPresent(Int.self, forKey: .expiresInSeconds)
        tokenType = try container.decodeIfPresent(String.self, forKey: .tokenType) ?? "Bearer"
    }

    public func encode(to encoder: Encoder) throws {
        var container = encoder.container(keyedBy: CodingKeys.self)
        try container.encode(accessToken, forKey: .accessToken)
        try container.encodeIfPresent(refreshToken, forKey: .refreshToken)
        try container.encodeIfPresent(expiresInSeconds, forKey: .expiresInSeconds)
        try container.encode(tokenType, forKey: .tokenType)
    }
}

public enum GSXAuthStrategy: Equatable {
    case password(email: String, password: String)
    case oauth(provider: String, code: String, redirectURI: String? = nil, codeVerifier: String? = nil)
    case webAuthn(challengeID: String, clientDataJSON: String, authenticatorData: String, signature: String, userHandle: String? = nil)
    case custom(strategy: String, credentials: [String: String] = [:], attributes: [String: String] = [:])

    public var request: GSXAuthExchangeRequest {
        switch self {
        case .password(let email, let password):
            return GSXAuthExchangeRequest(
                strategy: "password",
                credentials: ["email": email, "password": password]
            )
        case .oauth(let provider, let code, let redirectURI, let codeVerifier):
            var credentials = ["provider": provider, "code": code]
            if let redirectURI, !redirectURI.isEmpty {
                credentials["redirect_uri"] = redirectURI
            }
            if let codeVerifier, !codeVerifier.isEmpty {
                credentials["code_verifier"] = codeVerifier
            }
            return GSXAuthExchangeRequest(strategy: "oauth", credentials: credentials)
        case .webAuthn(let challengeID, let clientDataJSON, let authenticatorData, let signature, let userHandle):
            var credentials = [
                "challenge_id": challengeID,
                "client_data_json": clientDataJSON,
                "authenticator_data": authenticatorData,
                "signature": signature,
            ]
            if let userHandle, !userHandle.isEmpty {
                credentials["user_handle"] = userHandle
            }
            return GSXAuthExchangeRequest(strategy: "webauthn", credentials: credentials)
        case .custom(let strategy, let credentials, let attributes):
            return GSXAuthExchangeRequest(strategy: strategy, credentials: credentials, attributes: attributes)
        }
    }
}

public final class GSXAuthClient {
    private let dataClient: GSXDataClient

    public init(dataClient: GSXDataClient) {
        self.dataClient = dataClient
    }

    public convenience init(transport: any GSXTransport) {
        self.init(dataClient: GSXDataClient(transport: transport))
    }

    public convenience init(baseURL: URL, defaultHeaders: [String: String] = [:]) {
        self.init(dataClient: GSXDataClient(baseURL: baseURL, defaultHeaders: defaultHeaders))
    }

    public convenience init(baseURL: String, defaultHeaders: [String: String] = [:]) throws {
        self.init(dataClient: try GSXDataClient(baseURL: baseURL, defaultHeaders: defaultHeaders))
    }

    public func exchange(
        _ request: GSXAuthExchangeRequest,
        path: String = "/api/auth/exchange",
        policy: GSXRequestPolicy = GSXRequestPolicy(name: "auth.exchange", auth: .none, retryAttempts: 1)
    ) async throws -> GSXAuthTokenResponse {
        let response = try await dataClient.submit(
            try GSXRequest.json(method: "POST", path: path, body: request),
            policy: policy
        )
        return try response.decodedJSON(GSXAuthTokenResponse.self)
    }

    public func exchange(
        _ strategy: GSXAuthStrategy,
        path: String = "/api/auth/exchange",
        policy: GSXRequestPolicy = GSXRequestPolicy(name: "auth.exchange", auth: .none, retryAttempts: 1)
    ) async throws -> GSXAuthTokenResponse {
        try await exchange(strategy.request, path: path, policy: policy)
    }

    public func exchange(
        strategy: String,
        credentials: [String: String] = [:],
        attributes: [String: String] = [:],
        path: String = "/api/auth/exchange",
        policy: GSXRequestPolicy = GSXRequestPolicy(name: "auth.exchange", auth: .none, retryAttempts: 1)
    ) async throws -> GSXAuthTokenResponse {
        try await exchange(
            GSXAuthExchangeRequest(strategy: strategy, credentials: credentials, attributes: attributes),
            path: path,
            policy: policy
        )
    }

    public func signInWithPassword(
        email: String,
        password: String,
        path: String = "/api/auth/exchange",
        policy: GSXRequestPolicy = GSXRequestPolicy(name: "auth.exchange", auth: .none, retryAttempts: 1)
    ) async throws -> GSXAuthTokenResponse {
        try await exchange(.password(email: email, password: password), path: path, policy: policy)
    }

    public func signInWithOAuth(
        provider: String,
        code: String,
        redirectURI: String? = nil,
        codeVerifier: String? = nil,
        path: String = "/api/auth/exchange",
        policy: GSXRequestPolicy = GSXRequestPolicy(name: "auth.exchange", auth: .none, retryAttempts: 1)
    ) async throws -> GSXAuthTokenResponse {
        try await exchange(.oauth(provider: provider, code: code, redirectURI: redirectURI, codeVerifier: codeVerifier), path: path, policy: policy)
    }

    public func signInWithWebAuthn(
        challengeID: String,
        clientDataJSON: String,
        authenticatorData: String,
        signature: String,
        userHandle: String? = nil,
        path: String = "/api/auth/exchange",
        policy: GSXRequestPolicy = GSXRequestPolicy(name: "auth.exchange", auth: .none, retryAttempts: 1)
    ) async throws -> GSXAuthTokenResponse {
        try await exchange(
            .webAuthn(
                challengeID: challengeID,
                clientDataJSON: clientDataJSON,
                authenticatorData: authenticatorData,
                signature: signature,
                userHandle: userHandle
            ),
            path: path,
            policy: policy
        )
    }
}

public final class GSXAuthSession {
    private let client: GSXAuthClient
    private let tokenStore: any GSXMutableTokenStore

    public init(client: GSXAuthClient, tokenStore: any GSXMutableTokenStore) {
        self.client = client
        self.tokenStore = tokenStore
    }

    public convenience init(transport: any GSXTransport, tokenStore: any GSXMutableTokenStore) {
        self.init(client: GSXAuthClient(transport: transport), tokenStore: tokenStore)
    }

    public convenience init(baseURL: URL, defaultHeaders: [String: String] = [:], tokenStore: any GSXMutableTokenStore) {
        self.init(client: GSXAuthClient(baseURL: baseURL, defaultHeaders: defaultHeaders), tokenStore: tokenStore)
    }

    public convenience init(baseURL: String, defaultHeaders: [String: String] = [:], tokenStore: any GSXMutableTokenStore) throws {
        self.init(client: try GSXAuthClient(baseURL: baseURL, defaultHeaders: defaultHeaders), tokenStore: tokenStore)
    }

    @discardableResult
    public func signIn(
        _ strategy: GSXAuthStrategy,
        path: String = "/api/auth/exchange",
        policy: GSXRequestPolicy = GSXRequestPolicy(name: "auth.exchange", auth: .none, retryAttempts: 1)
    ) async throws -> GSXAuthTokenResponse {
        let response = try await client.exchange(strategy, path: path, policy: policy)
        try await tokenStore.setToken(response.accessToken)
        return response
    }

    public func signOut() async throws {
        try await tokenStore.clearToken()
    }
}

File: runtime/android/src/main/kotlin/com/gosx/nativekit/GSXAuth.kt
package com.gosx.nativekit

import org.json.JSONObject

data class GSXAuthExchangeRequest(
    val strategy: String,
    val credentials: Map<String, String> = emptyMap(),
    val attributes: Map<String, String> = emptyMap(),
) {
    fun toJSON(): String {
        val objectValue = JSONObject()
        objectValue.put("strategy", strategy)
        objectValue.put("credentials", JSONObject(credentials))
        objectValue.put("attributes", JSONObject(attributes))
        return objectValue.toString()
    }
}

data class GSXAuthTokenResponse(
    val accessToken: String,
    val refreshToken: String? = null,
    val expiresInSeconds: Int? = null,
    val tokenType: String = "Bearer",
) {
    companion object {
        fun fromJSON(json: String): GSXAuthTokenResponse {
            val objectValue = JSONObject(json)
            return GSXAuthTokenResponse(
                accessToken = objectValue.getString("access_token"),
                refreshToken = objectValue.optString("refresh_token").ifBlank { null },
                expiresInSeconds = if (objectValue.has("expires_in") && !objectValue.isNull("expires_in")) {
                    objectValue.getInt("expires_in")
                } else {
                    null
                },
                tokenType = objectValue.optString("token_type", "Bearer").ifBlank { "Bearer" },
            )
        }
    }
}

sealed class GSXAuthStrategy {
    abstract fun request(): GSXAuthExchangeRequest

    data class Password(
        val email: String,
        val password: String,
    ) : GSXAuthStrategy() {
        override fun request(): GSXAuthExchangeRequest = GSXAuthExchangeRequest(
            strategy = "password",
            credentials = mapOf("email" to email, "password" to password),
        )
    }

    data class OAuth(
        val provider: String,
        val code: String,
        val redirectURI: String? = null,
        val codeVerifier: String? = null,
    ) : GSXAuthStrategy() {
        override fun request(): GSXAuthExchangeRequest {
            val credentials = mutableMapOf("provider" to provider, "code" to code)
            if (!redirectURI.isNullOrBlank()) {
                credentials["redirect_uri"] = redirectURI
            }
            if (!codeVerifier.isNullOrBlank()) {
                credentials["code_verifier"] = codeVerifier
            }
            return GSXAuthExchangeRequest(strategy = "oauth", credentials = credentials)
        }
    }

    data class WebAuthn(
        val challengeID: String,
        val clientDataJSON: String,
        val authenticatorData: String,
        val signature: String,
        val userHandle: String? = null,
    ) : GSXAuthStrategy() {
        override fun request(): GSXAuthExchangeRequest {
            val credentials = mutableMapOf(
                "challenge_id" to challengeID,
                "client_data_json" to clientDataJSON,
                "authenticator_data" to authenticatorData,
                "signature" to signature,
            )
            if (!userHandle.isNullOrBlank()) {
                credentials["user_handle"] = userHandle
            }
            return GSXAuthExchangeRequest(strategy = "webauthn", credentials = credentials)
        }
    }

    data class Custom(
        val strategy: String,
        val credentials: Map<String, String> = emptyMap(),
        val attributes: Map<String, String> = emptyMap(),
    ) : GSXAuthStrategy() {
        override fun request(): GSXAuthExchangeRequest = GSXAuthExchangeRequest(
            strategy = strategy,
            credentials = credentials,
            attributes = attributes,
        )
    }
}

class GSXAuthClient(
    private val dataClient: GSXDataClient,
) {
    constructor(transport: GSXTransport) : this(GSXDataClient(transport = transport))

    constructor(baseURL: String, defaultHeaders: Map<String, String> = emptyMap()) : this(
        GSXDataClient(baseURL = baseURL, defaultHeaders = defaultHeaders),
    )

    suspend fun exchange(
        request: GSXAuthExchangeRequest,
        path: String = "/api/auth/exchange",
        policy: GSXRequestPolicy = GSXRequestPolicy(name = "auth.exchange", auth = GSXAuthRequirement.None, retryAttempts = 1),
    ): GSXAuthTokenResponse {
        val response = dataClient.submit(
            GSXRequest.json(method = "POST", path = path, json = request.toJSON()),
            policy = policy,
        )
        return GSXAuthTokenResponse.fromJSON(response.text())
    }

    suspend fun exchange(
        strategy: GSXAuthStrategy,
        path: String = "/api/auth/exchange",
        policy: GSXRequestPolicy = GSXRequestPolicy(name = "auth.exchange", auth = GSXAuthRequirement.None, retryAttempts = 1),
    ): GSXAuthTokenResponse =
        exchange(request = strategy.request(), path = path, policy = policy)

    suspend fun exchange(
        strategy: String,
        credentials: Map<String, String> = emptyMap(),
        attributes: Map<String, String> = emptyMap(),
        path: String = "/api/auth/exchange",
        policy: GSXRequestPolicy = GSXRequestPolicy(name = "auth.exchange", auth = GSXAuthRequirement.None, retryAttempts = 1),
    ): GSXAuthTokenResponse =
        exchange(
            request = GSXAuthExchangeRequest(strategy = strategy, credentials = credentials, attributes = attributes),
            path = path,
            policy = policy,
        )

    suspend fun signInWithPassword(
        email: String,
        password: String,
        path: String = "/api/auth/exchange",
        policy: GSXRequestPolicy = GSXRequestPolicy(name = "auth.exchange", auth = GSXAuthRequirement.None, retryAttempts = 1),
    ): GSXAuthTokenResponse =
        exchange(strategy = GSXAuthStrategy.Password(email = email, password = password), path = path, policy = policy)

    suspend fun signInWithOAuth(
        provider: String,
        code: String,
        redirectURI: String? = null,
        codeVerifier: String? = null,
        path: String = "/api/auth/exchange",
        policy: GSXRequestPolicy = GSXRequestPolicy(name = "auth.exchange", auth = GSXAuthRequirement.None, retryAttempts = 1),
    ): GSXAuthTokenResponse =
        exchange(
            strategy = GSXAuthStrategy.OAuth(
                provider = provider,
                code = code,
                redirectURI = redirectURI,
                codeVerifier = codeVerifier,
            ),
            path = path,
            policy = policy,
        )

    suspend fun signInWithWebAuthn(
        challengeID: String,
        clientDataJSON: String,
        authenticatorData: String,
        signature: String,
        userHandle: String? = null,
        path: String = "/api/auth/exchange",
        policy: GSXRequestPolicy = GSXRequestPolicy(name = "auth.exchange", auth = GSXAuthRequirement.None, retryAttempts = 1),
    ): GSXAuthTokenResponse =
        exchange(
            strategy = GSXAuthStrategy.WebAuthn(
                challengeID = challengeID,
                clientDataJSON = clientDataJSON,
                authenticatorData = authenticatorData,
                signature = signature,
                userHandle = userHandle,
            ),
            path = path,
            policy = policy,
        )
}

class GSXAuthSession(
    private val client: GSXAuthClient,
    private val tokenStore: GSXMutableTokenStore,
) {
    constructor(transport: GSXTransport, tokenStore: GSXMutableTokenStore) : this(
        client = GSXAuthClient(transport = transport),
        tokenStore = tokenStore,
    )

    constructor(
        baseURL: String,
        defaultHeaders: Map<String, String> = emptyMap(),
        tokenStore: GSXMutableTokenStore,
    ) : this(
        client = GSXAuthClient(baseURL = baseURL, defaultHeaders = defaultHeaders),
        tokenStore = tokenStore,
    )

    suspend fun signIn(
        strategy: GSXAuthStrategy,
        path: String = "/api/auth/exchange",
        policy: GSXRequestPolicy = GSXRequestPolicy(name = "auth.exchange", auth = GSXAuthRequirement.None, retryAttempts = 1),
    ): GSXAuthTokenResponse {
        val response = client.exchange(strategy = strategy, path = path, policy = policy)
        tokenStore.setToken(response.accessToken)
        return response
    }

    suspend fun signOut() {
        tokenStore.clearToken()
    }
}
